doc/dvsource-jack.1
.\" dvsource-jack.1 written by Robin Gareus <robin@gareus.org>
.mso www.tmac
.TH DVSOURCE-JACK 1 "25 March 2011"
.SH NAME
dvsource-jack \- jack audio source for DVswitch
.SH SYNOPSIS
.HP
.B dvsource-jack
.RI [ OPTIONS ]
.RB [ \-s " ntsc|pal]
.SH DESCRIPTION
.LP
\fBdvsource-jack\fR uses JACK to capture audio and sends it to
DVswitch together with a dummy video stream.
.SH OPTIONS
\fB\-h\fR, \fB\-\-host=\fIHOST\fR
.TP
\fB\-p\fR, \fB\-\-port=\fIPORT\fR
.RS
Specify the network address on which DVswitch is listening.  The host
address may be specified by name or as an IPv4 or IPv6 literal.
.RE
.TP
\fB\-s\fR, \fB\-\-system=\fRntsc|pal
.RS
Specify the video system to use.  This must match the system used by
DVswitch.  The default is "pal".
.RE
.TP
\fB-d\fR, \fB--delay\fR=\fIDELAY\fR
.RS
Specify the time by which audio should be delayed, in seconds.
The delays in live audio and video sources need to be closely
matched in order that the audio and video signals will be
synchronised in the output of DVswitch.  The default is 0.2.
.RE
.SH AUTHOR
Robin Gareus <robin@gareus.org>.
.SH SEE ALSO
.UR http://jackaudio.org/
.UE ,
/usr/share/doc/dvswitch/README


ChangeLog
dvswitch (0.9~rc2) UNRELEASED; urgency=low

  [ Wouter Verhelst ]
  * Don't try to build or install dvsource-alsa on non-Linux.

  [ Robin Gareus ]
  * fast-fades (allow fade effect 40ms..15sec)
  * ffmpeg API update (for ffmpeg-0.10)
  * dynamic A/B overlay
  * confirm application quit dialog
  * blinking status bar
  * updated GUI widget alignment, and labels
  * various updates to Debian packaging
  * various spelling and formating fixes (doc, man-pages)

 -- Wouter Verhelst <wouter@debian.org>  Wed, 28 Mar 2012 00:39:04 +0200

dvswitch (0.9~rc1) experimental; urgency=low

  [ Ben Hutchings ]
  * Added dvsource-alsa as an audio-only source
  * Create one encoding thread per processor, up to a maximum of 8
  * Add status overlay to make it more clear whether recording is enabled
  * Set display to black before a source is connected
  * Add static Makefile at top level
  * Override source frame aspect ratio in case of a mismatch, rather then
    warning about it - Alioth feature #311953
  * Fix APTs and validity flags in header block - Alioth bug #311836
  * List all needed libraries to the linker.  We were accidentally
    depending on the behaviour of GNU ld, which treats 'needed' as
    transitive by default - Debian bug #554310
  * Remove support for 44.1 kHz audio
  * Add update-po target for updating translation files
  * Fix build failure with ffmpeg 0.6
  * Fix crash at startup with ffmpeg 0.6 due to a new parameter check
  * Enable selection of Firewire DV devices by GUID, thanks to
    Manuel Virgilio
  * Implement dvsource-firewire-rtsp, to serve DV over RTSP (experimental)

  [ Carl Karsten ]
  * Add peak level indicators to VU-meter

  [ Wouter Verhelst ]
  * Add support for tally lights in dvsource-firewire/dvsource-v4l2-dv
  * Internationalize and add Dutch translation
  * Disable effects buttons until sufficient sources have connected.
    Pressing them before sources are connected would result in a crash.
  * Implement fading on an 8-bit scale
  * Disable record button if we can't record - Alioth bug #312329
  * Add tally light support to dvsource-firewire-rtsp
  * Install dvsource-firewire-rtsp into .../bin
  * Take over day-to-day maintenance from Ben

  [ Robin Gareus ]
  * Add dvsource-jack, similar to dvsource-alsa but for JACK audio sources

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 01 Mar 2009 03:16:30 +0000

dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers
    - Changed selection of visual class and depth for images to match the
      window
    - Added error messages and later checks for failure of image creation,
      where previously dvswitch could crash

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 21 Feb 2008 22:17:32 +0000

dvswitch (0.6) unstable; urgency=low

  * Added some infrastructure for video effects including
    picture-in-picture mixing
  * Switched from gtkmm 2.0/2.2 to 2.4+
  * Corrected small error in video display width

src/dvsource-jack.c
/* Copyright 2007-2009 Ben Hutchings.
 * written 2011 by Robin Gareus <robin@gareus.org> 
 * See the file "COPYING" for licence details.
 */
/* Source that reads audio from JACK and combines with black video */

#include <assert.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>

#include <pthread.h>
#include <jack/jack.h>
#include <jack/ringbuffer.h>

#include "config.h"
#include "dif.h"
#include "pcm.h"
#include "protocol.h"
#include "socket.h"

static struct option options[] = {
    {"host",   1, NULL, 'h'},
    {"port",   1, NULL, 'p'},
    {"system", 1, NULL, 's'},
    {"delay",  1, NULL, 'd'},
    {"help",   0, NULL, 'H'},
    {NULL,     0, NULL, 0}
};

static char * mixer_host = NULL;
static char * mixer_port = NULL;
static int terminate = 0;

static void handle_config(const char * name, const char * value)
{
    if (strcmp(name, "MIXER_HOST") == 0)
    {
	free(mixer_host);
	mixer_host = strdup(value);
    }
    else if (strcmp(name, "MIXER_PORT") == 0)
    {
	free(mixer_port);
	mixer_port = strdup(value);
    }
}

static void usage(const char * progname)
{
    fprintf(stderr,
	    "\
Usage: %s [-h HOST] [-p PORT] [-s ntsc|pal] \\\n\
           [-d DELAY]\n",
	    progname);
}

struct transfer_params {
    jack_client_t                *j_client;
    jack_port_t                 **j_input_ports;
    jack_default_audio_sample_t **j_in;
    jack_ringbuffer_t            *rb;
    jack_nframes_t                j_sample_rate;
    unsigned int                  j_channel_count;

    pthread_t reader_thread_id;
    pthread_mutex_t reader_thread_lock;
    pthread_cond_t  buffer_ready;
    int             activated;
    long overruns; 

    jack_nframes_t          delay_size;

    const struct dv_system *system;
    enum dv_sample_rate     sample_rate_code;
    int                     sock;
};



/**
 * jack audio process callback
 */
int j_process (jack_nframes_t nframes, void *arg)
{
    jack_nframes_t c;
    struct transfer_params *params = (struct transfer_params*) arg;
    const unsigned int nports = params->j_channel_count;

    /* Do nothing until we're ready to begin. */
    if (!params->activated) return 0;

    for(c=0; c< nports; c++)
    {
	params->j_in[c] = (jack_default_audio_sample_t*) jack_port_get_buffer(params->j_input_ports[c], nframes);
    }
    //j_latency = jack_port_get_total_latency(params->j_client,j_input_port);


    /* interleave - 32 bit for now */
    jack_nframes_t s;
    for(s=0; s<nframes; s++)
    {
	for(c=0; c< nports; c++)
	{
	    if (jack_ringbuffer_write(params->rb, (char*) &params->j_in[c][s], sizeof(jack_default_audio_sample_t))
		< sizeof(jack_default_audio_sample_t))
	    {
		    params->overruns++;
		    break;
	    }
	}
    }
#if 1
    /* Tell the writer that there is work to do. */
    if(pthread_mutex_trylock(&params->reader_thread_lock) == 0)
    {
	pthread_cond_signal(&params->buffer_ready);
	pthread_mutex_unlock(&params->reader_thread_lock);
    }
#endif
  return 0;      
}

void close_jack(struct transfer_params * params)
{
    if (params->j_client) 
    {
	jack_deactivate(params->j_client);
	jack_client_close (params->j_client);
    }
    if (params->j_input_ports) free(params->j_input_ports);
    if (params->j_in) free(params->j_in);
    params->j_client = NULL;
    params->j_in = NULL;
    params->j_input_ports = NULL;
    terminate = 1;
}

void j_on_shutdown (void *arg)
{
  fprintf(stderr,"recv. shutdown request from jackd.\n");
  close_jack((struct transfer_params *) arg);
}


/**
 * open a client connection to the JACK server 
 */
int init_jack(struct transfer_params * params)
{
    // TODO use client_name etc, from params
    char * client_name = "dvsource";
    const char * server_name = NULL;
    jack_options_t options = JackNullOption;

    jack_status_t status;
    params->j_client = jack_client_open (client_name, options, &status, server_name);
    if (params->j_client == NULL)
    {
	fprintf (stderr, "jack_client_open() failed, status = 0x%2.0x\n", status);
	if (status & JackServerFailed)
	{
	    fprintf (stderr, "Unable to connect to JACK server\n");
	}
	return -1;
    }
    if (status & JackServerStarted)
    {
	fprintf (stderr, "JACK server started\n");
    }
    if (status & JackNameNotUnique)
    {
	client_name = jack_get_client_name(params->j_client);
	fprintf (stderr, "unique name `%s' assigned\n", client_name);
    }

    jack_set_process_callback (params->j_client, j_process, (void*) params);
    jack_on_shutdown (params->j_client, j_on_shutdown, (void*) params);
    params->j_sample_rate=jack_get_sample_rate (params->j_client);

    return 0;
}

/**
 *
 */
int jack_portsetup(struct transfer_params * params)
{
    jack_nframes_t i;

    const unsigned int nports = params->j_channel_count;
    const size_t in_size =  nports * sizeof (jack_default_audio_sample_t *);

    params->j_in = (jack_default_audio_sample_t **) malloc (in_size);
    params->j_input_ports = (jack_port_t **) malloc (sizeof (jack_port_t *) * nports);

    for (i=0; i < nports; i++)
    {
	char name[64];
	sprintf (name, "input%i", i+1);
	if ((params->j_input_ports[i] = jack_port_register (params->j_client, name, JACK_DEFAULT_AUDIO_TYPE, JackPortIsInput, 0)) == 0)
	{
	    fprintf (stderr, "cannot register input port \"%s\"!\n", name);
	    jack_client_close (params->j_client);
	    params->j_client=NULL;
	    return -1;
	}
    }
// TODO:  autoconnect
#if 0
    for (i = 0; i < nports; i++)
    {
	if (jack_connect (params->client, source_names[i], jack_port_name (params->j_input_ports[i])))
	{
	    fprintf (stderr, "cannot connect input port %s to %s\n", jack_port_name (params->j_input_ports[i]), source_names[i]);
	    jack_client_close (params->client);
	    params->j_client=NULL;
	    return (-1);
	} 
    }
#endif

    return 0;
}


int open_jack(struct transfer_params * params)
{
    if (init_jack(params))
    {
	close_jack(params);
	return -1;
    }
    if (jack_portsetup(params))
    {
	close_jack(params);
	return -1;
    }
    if (jack_activate (params->j_client))
    {
	close_jack(params);
	return -1;
    }
  return 0;
}

void *transfer_frames(void *arg)
{
    struct transfer_params *params = (struct transfer_params*) arg;
    static uint8_t buf[DIF_MAX_FRAME_SIZE];
    unsigned serial_num = 0;

    dv_buffer_fill_dummy(buf, params->system);

    pthread_setcanceltype (PTHREAD_CANCEL_ASYNCHRONOUS, NULL);
    pthread_mutex_lock (&params->reader_thread_lock);

    /* enable jack processing/buffering now. -- sync point */
    params->activated = 1;
    jack_nframes_t allocsize = 0;
    void *framebuf = NULL;

    while (!terminate)
    {
	const unsigned frame_count =
	    params->system->audio_frame_counts[params->sample_rate_code].std_cycle[
		serial_num % params->system->audio_frame_counts[params->sample_rate_code].std_cycle_len];

	const jack_nframes_t bytes_per_frame = frame_count * sizeof(jack_default_audio_sample_t) * params->j_channel_count;

	if (bytes_per_frame > allocsize)
	{
	    allocsize = bytes_per_frame;
	    if (framebuf) free(framebuf);
	    framebuf = malloc (bytes_per_frame);
	    printf("[re] allocated framebuffer: %lu bytes \n", (long int)bytes_per_frame);
	}

	while (params->activated &&
		   (jack_ringbuffer_read_space (params->rb) >= bytes_per_frame + params->delay_size))
	{

	    jack_ringbuffer_read (params->rb, framebuf, bytes_per_frame);
	    dv_buffer_set_audio(buf, params->sample_rate_code, frame_count, framebuf);

	    if (write(params->sock, buf, params->system->size)
		!= (ssize_t)params->system->size)
	    {
		perror("ERROR: write");
		exit(1);
	    }
	    ++serial_num;
	}

	/* wait until process() signals more data */
	pthread_cond_wait (&params->buffer_ready, &params->reader_thread_lock);
    }
    close(params->sock);
    params->activated = 0;
    free (framebuf);
    pthread_mutex_unlock(&params->reader_thread_lock);
    return (NULL);
}

int main(int argc, char ** argv)
{
    /* Initialise settings from configuration files. */
    dvswitch_read_config(handle_config);
    char * system_name = NULL;
    double delay = 0.2;

    struct transfer_params params;

    /* init and default params */
    pthread_mutex_init(&params.reader_thread_lock, NULL);
    pthread_cond_init(&params.buffer_ready, NULL);
    params.activated = params.overruns = 0;
    params.j_client = NULL; 
    params.rb = NULL; 
    params.j_in = NULL; 
    params.j_input_ports = NULL;
    params.j_channel_count = 2; /* hardcoded stereo */

    /* Parse arguments. */
    int opt;
    while ((opt = getopt_long(argc, argv, "h:p:s:r:d:", options, NULL)) != -1)
    {
	switch (opt)
	{
	case 'h':
	    free(mixer_host);
	    mixer_host = strdup(optarg);
	    break;
	case 'p':
	    free(mixer_port);
	    mixer_port = strdup(optarg);
	    break;
	case 's':
	    free(system_name);
	    system_name = strdup(optarg);
	    break;
	case 'd':
	    delay = strtod(optarg, NULL);
	    break;
	case 'H': /* --help */
	    usage(argv[0]);
	    return 0;
	default:
	    usage(argv[0]);
	    return 2;
	}
    }

    if (!mixer_host || !mixer_port)
    {
	fprintf(stderr, "%s: mixer hostname and port not defined\n",
		argv[0]);
	return 2;
    }

    if (!system_name || !strcasecmp(system_name, "pal"))
    {
	params.system = &dv_system_625_50;
    }
    else if (!strcasecmp(system_name, "ntsc"))
    {
	params.system = &dv_system_525_60;
    }
    else
    {
	fprintf(stderr, "%s: invalid system name \"%s\"\n", argv[0], system_name);
	return 2;
    }

    if (open_jack(&params))
    {
	fprintf(stderr, "%s: can not connect to JACK\n", argv[0]);
	return 2;
    }

    if (params.j_sample_rate == 32000)
    {
	params.sample_rate_code = dv_sample_rate_32k;
    }
    else if (params.j_sample_rate == 48000)
    {
	params.sample_rate_code = dv_sample_rate_48k;
    }
    else
    {
	fprintf(stderr, "%s: invalid sample rate %ld (need 48k or 32k SPS)\n", argv[0], (long int) params.j_sample_rate);
	close_jack(&params);
	return 2;
    }

    if (delay >= 0.0)
    {
	params.delay_size = delay 
	    * params.j_sample_rate * params.j_channel_count
	    * sizeof(jack_default_audio_sample_t); 
    }
    else
    {
	fprintf(stderr, "%s: delay parameter must be positive.\n", argv[0]);
	close_jack(&params);
	return 2;
    }


    if (argc > optind)
    {
	close_jack(&params);
	fprintf(stderr, "%s: excess argument \"%s\"\n",
		argv[0], argv[optind + 1]);
	usage(argv[0]);
	return 2;
    }

    const size_t rbsize = params.delay_size  /* + 1 sec extra buffer: */
	+  params.j_sample_rate * params.j_channel_count
	   * sizeof(jack_default_audio_sample_t); 
    params.rb = jack_ringbuffer_create(rbsize);
    memset(params.rb->buf, 0, rbsize);


    printf("INFO: Connecting to %s:%s\n", mixer_host, mixer_port);
    params.sock = create_connected_socket(mixer_host, mixer_port);
    assert(params.sock >= 0); /* create_connected_socket() should handle errors */
    if (write(params.sock, GREETING_SOURCE, GREETING_SIZE) != GREETING_SIZE)
    {
	close_jack(&params);
	perror("ERROR: write");
	exit(1);
    }
    printf("INFO: Connected.\n");

    pthread_create(&params.reader_thread_id, NULL, transfer_frames, &params);
    pthread_yield();

    /* TODO: interactively wait here (inc/dec delay, quit, etc) */
    while (!terminate) sleep (1);

    /* terminate and clean up*/

    close_jack(&params);

    if(params.activated)
    {
	terminate = 1;
	if(pthread_mutex_trylock(&params.reader_thread_lock) == 0)
	{
	    pthread_cond_signal(&params.buffer_ready);
	    pthread_mutex_unlock(&params.reader_thread_lock);
	}
	pthread_join(params.reader_thread_id, NULL);
    }

    //close(params.sock);
    if (params.rb) jack_ringbuffer_free(params.rb);

    printf("bye. and BTW: there were %li buffer overruns\n", params.overruns);

    exit(0);
    return 0;
}
/* vim: set sw=4 ts=8 sts=4: */
